### ISSUE TITLE
[cranelift library] panic! triggered in Solver::reassign_in (regalloc/solver.rs)

### ISSUE THREAD

**Issue by pventuzelo** [wasmerio/wasmer#728]
Thanks for the bug report!

### Describe the bug

`panic!` is called (line 558) in `cranelift/cranelift-codegen/src/regalloc/solver.rs`
This issue is making wasmer to panicked.

https://github.com/CraneStation/cranelift/blob/8222380a380cd6632012c83d14901572bda26298/cranelift-codegen/src/regalloc/solver.rs#L557-L560

### Expected behavior

`cranelift` should generate an exception instead.

This exception can be correctly handle by `wasmer` with the current code.  

https://github.com/wasmerio/wasmer/blob/77fe15db31b6bda5fca91698109a3712f35ea74c/lib/clif-backend/src/code.rs#L304-L305

### Status of my environment

wasmer version: commit 854b2013ea6f52d5d4d9a145e6607e5fa0cf573b

```bash
echo "`./target/release/wasmer -V` | `rustc -V` | `uname -m`"
wasmer 0.6.0 | rustc 1.38.0-nightly (ad7c55e1f 2019-08-07) | x86_64
```

### Steps to reproduce

Download 
[panic_cranelift.zip](https://github.com/wasmerio/wasmer/files/3547336/panic_cranelift.zip)

```
$ unzip panic_cranelift.zip
$ ./target/release/wasmer run panic_cranelift.wasm
thread 'main' panicked at 'Invalid from register for fixed v12 constraint', /home/scop/.cargo/registry/src/github.com-1ecc6299db9ec823/cranelift-codegen-0.31.0/src/regalloc/solver.rs:558:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

```

### Actual behavior

```
$ RUST_BACKTRACE=1 ./target/release/wasmer run panic_cranelift.wasm
thread 'main' panicked at 'Invalid from register for fixed v12 constraint', /home/scop/.cargo/registry/src/github.com-1ecc6299db9ec823/cranelift-codegen-0.31.0/src/regalloc/solver.rs:558:17
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.34/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.34/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:47
   3: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:36
   4: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:200
   5: std::panicking::default_hook
             at src/libstd/panicking.rs:214
   6: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:477
   7: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:384
   8: std::panicking::begin_panic_fmt
             at src/libstd/panicking.rs:339
   9: cranelift_codegen::regalloc::solver::Solver::reassign_in
  10: cranelift_codegen::regalloc::coloring::Context::visit_inst
  11: cranelift_codegen::regalloc::context::Context::run
  12: cranelift_codegen::context::Context::compile
  13: cranelift_codegen::context::Context::compile_and_emit
  14: wasmer_clif_backend::resolver::FuncResolverBuilder::new::{{closure}}
  15: <core::iter::adapters::Map<I,F> as core::iter::traits::iterator::Iterator>::try_fold::{{closure}}
  16: <core::iter::adapters::Map<I,F> as core::iter::traits::iterator::Iterator>::try_fold
  17: <rayon::iter::fold::FoldFolder<C,ID,F> as rayon::iter::plumbing::Folder<T>>::consume_iter
  18: <rayon::iter::while_some::WhileSomeFolder<C> as rayon::iter::plumbing::Folder<core::option::Option<T>>>::consume_iter
  19: rayon::iter::plumbing::bridge_producer_consumer::helper
  20: <rayon::iter::map::Map<I,F> as rayon::iter::ParallelIterator>::drive_unindexed
  21: rayon::iter::collect::<impl rayon::iter::ParallelExtend<T> for alloc::vec::Vec<T>>::par_extend
  22: rayon::result::<impl rayon::iter::FromParallelIterator<core::result::Result<T,E>> for core::result::Result<C,E>>::from_par_iter
  23: wasmer_clif_backend::resolver::FuncResolverBuilder::new
  24: <wasmer_clif_backend::code::CraneliftModuleCodeGenerator as wasmer_runtime_core::codegen::ModuleCodeGenerator<wasmer_clif_backend::code::CraneliftFunctionCodeGenerator,wasmer_clif_backend::signal::Caller,wasmer_clif_backend::code::CodegenError>>::finalize
  25: <wasmer_runtime_core::codegen::StreamingCompiler<MCG,FCG,RM,E,CGEN> as wasmer_runtime_core::backend::Compiler>::compile
  26: wasmer_runtime_core::compile_with_config
  27: wasmer_runtime::compile_with_config_with
  28: wasmer::execute_wasm
  29: wasmer::main
  30: std::rt::lang_start::{{closure}}
  31: std::rt::lang_start_internal::{{closure}}
             at src/libstd/rt.rs:49
  32: std::panicking::try::do_call
             at src/libstd/panicking.rs:296
  33: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:80
  34: std::panicking::try
             at src/libstd/panicking.rs:275
  35: std::panic::catch_unwind
             at src/libstd/panic.rs:394
  36: std::rt::lang_start_internal
             at src/libstd/rt.rs:48
  37: main
  38: __libc_start_main
  39: _start
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by pventuzelo** [comment]
TODO: tag with fuzzer-trophy

**Comment by pventuzelo** [comment]
Related  cranelift issue: https://github.com/CraneStation/cranelift/issues/947

**Comment by pventuzelo** [comment]
I've done multiple tests to figure what is the issue there.
I have extract the cranelift IR that we generate and potentially figure one additional issue but not related to this one actually.

When running the crashing sample in `cranelift v0.31` and `v0.41`. Everything work well on there side so i suppose it's maybe related to our version of cranelift.

Quick question: Why do we have [our own version of cranelift](https://github.com/wasmerio/cranelift)?

If you wasnt to reproduce the tests with `cranelift`:
``` sh 
cd cranelift
cargo run --release compile -c bug.clif --target x86_64
```

bug.clif is the following :
```
target x86_64

set opt_level=best
set baldrdash_prologue_words=0
set probestack_size_log2=12
set enable_verifier=false
set is_pic=false
set colocated_libcalls=false
set avoid_div_traps=false
set enable_float=true
set enable_nan_canonicalization=false
set enable_simd=true
set enable_atomics=true
set allones_funcaddrs=false
set probestack_enabled=true
set probestack_func_adjusts_sp=false
set jump_tables_enabled=false

target x86_64
target x86_64 has_sse3=false
target x86_64 has_ssse3=false
target x86_64 has_sse41=false
target x86_64 has_sse42=false
target x86_64 has_popcnt=false
target x86_64 has_avx=false
target x86_64 has_bmi1=false
target x86_64 has_bmi2=false
target x86_64 has_lzcnt=false


function u0:0(i64 vmctx, i32) -> i32 system_v {
    gv0 = vmctx
    gv1 = load.i64 notrap aligned readonly gv0+24
    gv2 = iadd_imm.i64 gv1, 0
    gv3 = load.i64 notrap aligned readonly gv2
    gv4 = load.i64 notrap aligned gv3
    gv5 = load.i32 notrap aligned gv3+8
    heap0 = dynamic gv4, min 0x0001_0000, bound gv5, offset_guard 4096, index_type i32

ebb0(v0: i64, v1: i32):
    v3 = iconst.i32 0
    v4 = iconst.i32 255
    v5 = heap_addr.i64 heap0, v4, 1
    v6 = uload8.i32 v5
    jump ebb2(v6, v1, v3, v3)

ebb2(v7: i32, v32: i32, v34: i32, v66: i32):
    v18 -> v7
    v36 -> v7
    v37 -> v18
    v31 -> v32
    v64 -> v66
    v65 -> v66
    v67 -> v64
    v8 = iconst.i32 -2
    v9 = udiv v7, v8
    v10 = iconst.i32 -7041
    brz v10, ebb5
    v11 = iconst.i32 0
    jump ebb4(v11)

ebb5:
    brnz.i32 v9, ebb4(v34)
    trap unreachable

ebb4(v33: i32):
    v15 -> v33
    v88 -> v33
    v35 -> v15
    jump ebb7(v31)

ebb7(v12: i32):
    v25 -> v12
    v68 -> v12
    v69 -> v12
    v13 = heap_addr.i64 heap0, v12, 1
    v14 = uload8.i32 v13+8
    v16 = icmp eq v14, v15
    v17 = bint.i32 v16
    brnz v17, ebb6
    v19 = icmp_imm.i32 eq v18, 0
    v20 = bint.i32 v19
    brz v20, ebb9
    v21 = iconst.i32 46
    v22 = icmp eq v12, v21
    v23 = bint.i32 v22
    brz v23, ebb10
    v24 = iconst.i32 1
    return v24

ebb10:
    jump ebb9

ebb9:
    v26 = heap_addr.i64 heap0, v12, 1
    v27 = load.i32 v26
    v28 = icmp_imm eq v27, 0
    v29 = bint.i32 v28
    brz v29, ebb11
    v30 = iconst.i32 14
    trap unreachable

ebb11:
    jump ebb7(v27)

ebb6:
    v38 = icmp_imm.i32 eq v15, 0
    v39 = bint.i32 v38
    brz v39, ebb12
    v40 = iconst.i32 1
    return v40

ebb12:
    v41 = heap_addr.i64 heap0, v12, 1
    v42 = uload8.i32 v41+9
    v43 = icmp_imm eq v42, 0
    v44 = bint.i32 v43
    brz v44, ebb14
    jump ebb13(v18, v15, v66)

ebb14:
    v45 = icmp.i32 ugt v42, v18
    v46 = bint.i32 v45
    brz v46, ebb15
    v47 = iconst.i32 2
    return v47

ebb15:
    v48 = iconst.i32 10
    v49 = udiv.i32 v12, v48
    v50 = ushr v49, v12
    v61 -> v50
    jump ebb17(v18, v66)

ebb17(v51: i32, v55: i32):
    v52 = iconst.i32 -2012
    v53 = heap_addr.i64 heap0, v52, 1
    v54 = uload8.i32 v53
    v56 = heap_addr.i64 heap0, v55, 1
    v57 = uload8.i32 v56
    v58 = sdiv v54, v57
    brz v58, ebb19
    v59 = iconst.i32 0
    return v59

ebb19:
    v60 = iconst.i32 -2775
    v62 = icmp eq v60, v61
    v63 = bint.i32 v62
    brnz v63, ebb16
    jump ebb17(v52, v60)

ebb16:
    jump ebb13(v52, v33, v60)

ebb13(v74: i32, v87: i32, v89: i32):
    v70 = heap_addr.i64 heap0, v12, 1
    v71 = load.i32 v70+4
    v72 = icmp_imm eq v71, 0
    v73 = bint.i32 v72
    brz v73, ebb20
    v75 = icmp_imm eq v74, 0
    v76 = bint.i32 v75
    brz v76, ebb21
    v77 = iconst.i32 1
    return v77

ebb21:
    v78 = iconst.i32 -1
    v79 = iadd.i32 v74, v78
    v80 = heap_addr.i64 heap0, v79, 1
    v81 = uload8.i32 v80
    v82 = iconst.i32 46
    v83 = icmp eq v81, v82
    v84 = bint.i32 v83
    brz v84, ebb22
    v85 = iconst.i32 2
    return v85

ebb22:
    v86 = iconst.i32 0
    return v86

ebb20:
    jump ebb2(v74, v71, v87, v89)
}
```

**Comment by MarkMcCaskey** [comment]
> Quick question: Why do we have our own version of cranelift?

It's an unfortunate side-effect of the fact that we made some changes that were unlikely to be accepted upstream due to non-technical reasons and that we haven't been vigilant about documenting these changes or keeping them organized.

I spent a good amount of time trying to fix this, but it wasn't obvious to me why we made certain changes and I couldn't get answers from anyone else.  It's an on-going issue that we'll have to deal with.  We definitely would much rather not have this fork
